rest-server/core/src/main/scala/com/actionml/core/jobs/JobManager.scala

```diff
@@ -45,87 +46,78 @@ trait JobManagerInterface {
-    val description = JobDescription(createUUID, JobStatus.executing, comment)
+  override def addJob(engineId: String, f: Future[_], c: Cancellable, comment: String): JobDescription = {
+    val description = JobDescription(createUUID, JobStatuses.executing, comment)
     val newJobDescriptions = jobDescriptions.getOrElse(engineId, Map.empty) +
-      (description.jobId -> (c -> description.copy(status = JobStatus.executing)))
-    jobDescriptions = jobDescriptions + (engineId -> newJobDescriptions)
+      (description.jobId -> (c -> description.copy(status = JobStatuses.executing)))
+    jobDescriptions.put(engineId, newJobDescriptions)
     f.onComplete {
       case Success(_) =>
         logger.info(s"Job ${description.jobId} completed successfully [engine $engineId]")
         removeJob(description.jobId)
       case Failure(e) =>
         logger.error(s"Job $description failed [engine $engineId]", e)
-        removeJob(description.jobId)
+        cancelJob(description.jobId)
     }
     description
   }
 
-  private def createUUID: String = UUID.randomUUID().toString
-
   /** Gets any active Jobs for the specified Engine */
   override def getActiveJobDescriptions(engineId: String): Map[String, JobDescription] = {
-    jobDescriptions.getOrElse(engineId, Map.empty).map {
-      case Tuple2(id, Tuple2(c, d)) => id -> d
-    }
+    (jobDescriptions.getOrElse(engineId, Map.empty).map {
+      case Tuple2(id, Tuple2(_, d)) => d
+    } ++ LivyJobServerSupport.status(engineId)).map(d => d.jobId -> d).toMap
   }
 
   override def removeJob(harnessJobId: String): Unit = {
-    jobDescriptions = jobDescriptions.map { case (engineId, jds) =>
+    jobDescriptions.map { case (engineId, jds) =>
       engineId -> (jds - harnessJobId)
     }
   }
 
   override def cancelJob(jobId: String): Future[Unit] = {
-    jobDescriptions.foreach {
+    Future.sequence(jobDescriptions.flatMap {
       case (_, jds) if jds.contains(jobId) =>
-        jds.get(jobId).foreach {
+        jds.get(jobId).map {
           case (cancellable, _) =>
             removeJob(jobId)
             cancellable.cancel()
         }
-    }
-    Future.successful(())
+    } ++ Seq(LivyJobServerSupport.cancel(jobId)))
+      .map(_ => ())
+      .recover {
+        case e: Exception =>
+          logger.error("Cancel job error", e)
+      }
   }
 
-}
 
+  private def createUUID: String = UUID.randomUUID().toString
+}
 
 trait Cancellable {
   def cancel(): Future[Unit]
 }
+
 object Cancellable {
   val noop: Cancellable = new Cancellable {
     override def cancel() = Future.successful(())
   }
 }
 
-case class JobDescription(
-  jobId: String,
-  status: String = JobStatus.queued,
-  comment: String = "") extends Response
+case class JobDescription(jobId: String,
+                          status: JobStatus = JobStatuses.queued,
+                          comment: String = "") extends Response
```

Comment:
bad Scala style

Reply:
let's use scalafmt, IDEA makes that ^^^ every time....

Reply:
it is a setting you can fix in IDEA. I don't bother with lots of coding conventions but we should stick to the ones in the style docs, right?

See the scala style docs. I have to fix these by hand every time I find them. I'll let these through. If you need help I can screen shot my settings, which are not perfect but catch this error at least.